### ISSUE TITLE
Document key events' priority

### ISSUE THREAD

**Issue by hcpl** [gyscos/Cursive#151]
### Description

Cannot process a "modifier+bare key" key combination, only a "bare key".

### Configuration
```sh
$ uname -srm
Linux 4.2.0-42-generic x86_64
$ rustc --version
rustc 1.18.0 (03fc9d622 2017-06-06)
```

Cursive version 0.6.4

### Expected behaviour
Print on the upper-left corner:
- `ctrl+enter` on Ctrl+Enter;
- `shift+enter` on Shift+Enter;
- `alt+enter` on Alt+Enter.

### Actual behaviour
Nothing happens.

### Code to reproduce
```rust
extern crate cursive;

use cursive::{Cursive, Printer};
use cursive::event::{Event, EventResult, Key};
use cursive::vec::Vec2;
use cursive::view::View;


struct Foo {
    info: &'static str,
}

impl View for Foo {
    fn draw(&self, printer: &Printer) {
        printer.print((0, 0), self.info);
    }

    fn required_size(&mut self, constraint: Vec2) -> Vec2 {
        Vec2::from((self.info.len(), 1)).or_min(constraint)
    }

    fn on_event(&mut self, event: Event) -> EventResult {
        self.info = match event {
            //Event::Key(Key::Enter) => "enter",
            Event::Ctrl(Key::Enter) => "ctrl+enter",
            Event::Shift(Key::Enter) => "shift+enter",
            Event::Alt(Key::Enter) => "alt+enter",
            _ => return EventResult::Ignored,
        };

        EventResult::Consumed(None)
    }
}


fn main() {
    let mut siv = Cursive::new();
    siv.add_fullscreen_layer(Foo { info: "" });
    siv.run();
}
```

P.S. If the `Event::Key(Key::Enter) => "enter"` part is uncommented, the program will print `enter ` for all 4 cases.

**Comment by gyscos** [comment]
Hi, and thanks for the report!

The "Enter" key is a special one, and isn't very well supported with modifier keys. `Ctrl` and `Shift` don't seem to have any effect; though `Alt+Enter` should be detectable with some modification to the input system. If you try with another key, like `Home`, it should work better.

Now, there seems to be some problem detecting this modifier. Apparently `Ctrl-Home` is detected as `Ctrl-Shift-Home`, and some other confusion, so I'll need to have a look at that part. :S

**Comment by hcpl** [comment]
Oh, I missed the example, thanks for pointing out.

Since ncurses does not support new features of modern terminals I already knew that keyboard handling is messy, but I never thought it's _that_ messy. So I came up with a table:

|    | No modifiers | Shift | Ctrl | Alt | Ctrl+Shift | Ctrl+Alt | Alt+Shift | Ctrl+Alt+Shift |
|---|---|---|---|---|---|---|---|---|
| Enter | `Key(Enter)` | `Key(Enter)` | `Key(Enter)` | `Key(Esc)` `Key(Enter)` | `Key(Enter)` | `Key(Esc)` `Key(Enter)` | `Key(Esc)` `Key(Enter)` | `Key(Esc)` `Key(Enter)` |
| Home | `Key(Home)` | Nothing | `Key(Home)` | `Key(Home)` | Nothing | `Key(Home)` | Nothing | Nothing |
| End | `Key(End)` | Nothing | `Key(End)` | `Key(End)` | Nothing | `Key(End)` | Nothing | Nothing |
| PageUp | `Key(PageUp)` | Nothing | `CtrlAlt(PageDown)` | `Ctrl(PageDown)` | Nothing | `AltShift(PageUp)` | Nothing | Nothing |
| PageDown | `Key(PageDown)` | Nothing | `CtrlAlt(Left)` | `Ctrl(Left)` | Nothing | `AltShift(PageDown)` | Nothing | Nothing |
| Left | `Key(Left)` | `Shift(Left)` | `CtrlAlt(Ins)` | `Ctrl(Ins)` | `Alt(Left)` | `AltShift(Left)` | `Unknown([31, 2, 0, 0])` | `Key(Esc)` `Char('[')` `Char('1')` `Char(';')` `Char('8')` `Char('D')` |
| Right | `Key(Right)` | `Shift(Right)` | `CtrlAlt(PageUp)` | `Ctrl(PageUp)` | `Alt(Right)` | `AltShift(Right)` | `CtrlShift(PageUp)` | `Key(Esc)` `Char('[')` `Char('1')` `Char(';')` `Char('8')` `Char('C')` |
| Up | `Key(Up)` | `Shift(Up)` | `Unknown([53, 2, 0, 0])` | `CtrlShift(Right)` | Nothing | `AltShift(Up)` | `CtrlAlt(Right)` | Nothing |
| Down | `Key(Down)` | `Shift(Down)` | `Unknown([12, 2, 0, 0])` | `CtrlShift(Del)` | Nothing | `AltShift(Down)` | `Unknown([11, 2, 0, 0])` | Nothing |

**Comment by gyscos** [comment]
Another note: the terminal used has a large impact on the events actually delivered to the application.

I just pushed a1737ca0ba1a99bffc5ec6f42e3371c9df552a91, which at least fixes most modifier+key input on my computer for the ncurses backend. I tested it locally with gnome-terminal and konsole. Not all key combinations work on both, but I didn't see any `Unknown` or mislabeled event. If this work for you, I'll update the pancurses backend to work the same.

If this doesn't work for you, then it means the key codes are different between multiple systems (could be terminfo version, terminal used, ...), which would be a bit of a pain...

Notes:
* As said above, modifiers + Enter rarely work.
* Shift+Insert actually pastes the clipboard content in many applications. That said, Konsole at least can properly detect Alt+Shift+Insert. Gnome-terminal doesn't.
* PageUp/Down is often used for scrolling/changing tabs, especially with modifiers. On gnome-terminal, only Ctrl+Shift+PageUp/Down does not work. On konsole, in addition, Shift+PageUp/Down and Alt+Shift+PageUp/Down do not work.

For these reasons, I used to have in the Readme a compatibility table of the events available on various terminals, but it ended up being out-of-date, it depended on the backend used, and was overall too much work to maintain.

**Comment by hcpl** [comment]
Yeah, I only tested on gnome-terminal. I also have xterm and urxvt installed, so I'll test your commit against them too.

**Comment by gyscos** [comment]
After realizing that the actual key <=> code mapping was changing between ncurses versions, I understood hard-coding this table wasn't going anywhere. So now it should detect this mapping at runtime. This include the (`Alt`, `AltShift`, `Ctrl`, `CtrlShift` and `CtrlAlt`) modifiers, for the (`Del`, `Insert`, `Home`, End`, `PageUp`/`Down` and the arrow keys).

**Comment by kc8apf** [comment]
I'm seeing incorrect mapping with macOS Terminal and the default backend.  For example, Alt(Left) appears to be an escape code of Key(Esc) Char('b').  When I get some time, I'll dig into the code to see what can be done.

**Comment by xliiv** [comment]
What's the future of this task?
Is it doable?

**Comment by gyscos** [comment]
Input encoding depends on the terminal used, and those can vary a lot. Recognizing the encoding scheme and parsing the input appropriately is a huge task, and ncurses is supposed to do exactly that. As much as possible, I'd like to leave most of the `$TERM`-dependent work to the backend library (or, if needed, the backend module).

That being said, parsing the output from ncurses itself isn't always trivial; many key codes are correctly parsed, but the returned code isn't documented, and the actual value depends on the ncurses version. This is what can be improved here - for instance, confusing `CtrlAlt(Ins)` and `Ctrl(Left)` results from wrongly parsing ncurses' own output, and should be correctable. Receiving `Key(Esc)` and `Char('b')` instead of `Alt(Left)` means ncurses is not parsing the escape sequence for this particular terminal, and is trickier to fix here.

Regarding macOS compatibility, I don't have a machine to test; https://github.com/gyscos/Cursive/issues/206 may be of interest.

**Comment by massimo79m** [comment]
Same behaviour on macos terminal. i need the ctrl-enter but it does not work.
